File: AlgoritmosOrdenamiento/Comparaciones/src/MergeSort.cpp
#include <iostream>
#include <sstream> 
#include <fstream>
#include <ctime> 
#include <string> 

using namespace std;

unsigned t0, t1;

ofstream fs("Salidas/MergeSort/salida6cc.txt");

template < typename Type >
void Merge( Type *arr, int left, int mid, int right ) {

    Type *temp = new Type[ right - left + 1 ];
    int i = left;
    int j = mid + 1;
    int k = 0;

    while( i <= mid && j <= right ) {
        if( arr[ i ] <= arr[ j ])
            temp[ k++ ] = arr[ i++ ];
        else
            temp[ k++ ] = arr[ j++ ];
    }

    while( i <= mid )
        temp[ k++ ] = arr[ i++ ];
    while( j <= right )
        temp[ k++ ] = arr[ j++ ];
    for( k = 0, i = left; i <= right; ++i, ++k )
        arr[ i ] = temp[ k ];
 
    delete[] temp;
}

template < typename Type >
void MergeSort( Type *arr, int left, int right ) {
    
    int mid = ( left + right ) / 2;

    if( left < right ) {
        MergeSort( arr, left, mid );
        MergeSort( arr, mid + 1, right );
        Merge( arr, left, mid, right );
    }
}

template < typename Type >
void MergeSort( Type *arr, int size ) {
 
    MergeSort( arr, 0, size - 1 );
} 

int main(){
	int tam [] = {100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000, 1000000}; 

    for(int j=0;j<10;j++){
        int *A = new int[tam[j]];

        ifstream file;
        string num;
        int num_int=0;
        file.open("Entradas/aleatorio6.txt",ios::in);
        int i=0;
        while(!file.eof() && i<=tam[j]){
            getline(file,num);
            stringstream convert(num);
            convert>>num_int;
            A[i]=num_int;
            i++;
        }
        file.close();

        t0=clock();
        MergeSort(A,tam[j]);
        t1 = clock();

        double time = (double(t1-t0)/CLOCKS_PER_SEC);
        fs << tam[j] << " " << time <<endl;
        delete[] A;
    }

	return 0;

} 

File: AlgoritmosOrdenamiento/Comparaciones/src/MergeSort.py
from io import open
from time import time

def mergeSort(arr): 
	if len(arr) >1: 
		mid = len(arr)//2
		L = arr[:mid] 
		R = arr[mid:] 

		mergeSort(L) 
		mergeSort(R) 

		i = j = k = 0

		while i < len(L) and j < len(R): 
			if L[i] < R[j]: 
				arr[k] = L[i] 
				i+= 1
			else: 
				arr[k] = R[j] 
				j+= 1
			k+= 1

		while i < len(L): 
			arr[k] = L[i] 
			i+= 1
			k+= 1
		
		while j < len(R): 
			arr[k] = R[j] 
			j+= 1
			k+= 1



tam = [100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000, 1000000]
times = [0,0,0,0,0,0,0,0,0,0]
f=open("Salidas/MergeSort/salida1py.txt","w")

for j in range(10):            
    file = open("Entradas/aleatorio1.txt","r")
    A = file.readlines()
    file.close()

    for i in range(tam[j]):
        A[i]=int(A[i])

    t0=time()
    mergeSort(A) 
    tiempo = time()-t0
    tiempo = round(tiempo,3)

    A.clear()
    f.write(str(tam[j])+" "+str(tiempo)+"\n")
